### DISCUSSION TITLE
Basic example of how to add a node?

### DISCUSSION THREAD

**Discussion by NullVoxPopuli** [posthtml/posthtml#374] [Q&A]
I'm wanting to prepend a `link` tag before all existing `link` tags.
I've been reading the README, some tests, etc -- haven't found such a capability.

Is this possible? how would you do it?

**Comment by Scrum** [comment]
@NullVoxPopuli You can try this way, perhaps there is a more elegant solution

```js
const posthtml = require('posthtml')

const html = `<a>test1</a><span><a>test2</a></span><a>test1</a>`;
const plugins = [
  function(tree) {
    tree.walk(node => {
      if (node.tag === 'a' && (node.attrs === undefined || node.attrs.skip === undefined)) {
        node.attrs = {
          skip: true
        }

        return {
          tag: false,
          content: [
            {tag: 'a', content: 'test0', attrs: {skip: true}},
            node
          ]
        }
      }
      return node;
    });

    tree.match({tag: 'a'}, node => {
      delete node.attrs.skip;
      return node;
    })

    return tree;
  }
]
const options = {}

posthtml(plugins)
  .process(html, options)
  .then((result) =>  console.log(result.html))
```

**Reply by NullVoxPopuli** [reply]
what if you don't want to define an AST?
I imagine it works like this:
 - read in HTML document
 - insert `link` tag before all link tags (as string)
   - maybe the function I write parses this fragment to an AST and inserts that
 - write HTML document with no formatting changes

**Reply by NullVoxPopuli** [reply]
I made this REPL: https://runkit.com/nullvoxpopuli/posthtml---insertbeforefirst-link/1.0.0
_but_, it seems indentation is lost
![image](https://user-images.githubusercontent.com/199018/150619387-985ddd88-87db-499f-9d44-00f0328ebbb2.png)

**Reply by NullVoxPopuli** [reply]
I figured out a way to make it work -- but I'm not sure how scalable it is :thinking: 
https://runkit.com/nullvoxpopuli/posthtml---insertbeforefirst-link/2.0.0

**Reply by Scrum** [reply]
You can use the plugin https://github.com/posthtml/posthtml-beautify

**Comment by shoonia** [comment]
Have the same issue, I want to add a new node. Maybe add some method for creating a node to the core library? More "legal" way.

By now, I use a class. like this

```js
// Or better use a plain object?
class Node {
  constructor(tag, attrs, content) {
    this.tag = tag;
    this.attrs = attrs;
    this.content = content;
  }
}

posthtml().use((tree) => {
  tree.match({ tag: 'p' }, (node) => {
    node.content ??= [
        new Node('span', { class: 'some' }, ['text']),
    ];

    return node;
  })
});
```

**Reply by Scrum** [reply]
> Have the same issue, I want to add a new node. Maybe add some method for creating a node to the core library? More "legal" way.

The idea is interesting. I will think about how best to organize it and whether there is value in it.

https://github.com/posthtml/posthtml-parser/issues/88
